sklearn/metrics/classification.py

```diff
@@ -391,72 +392,266 @@ def jaccard_similarity_score(y_true, y_pred, normalize=True,
+                if not normalize:
+                    average = 'none-samples'
+            else:
+                raise ValueError("normalize != None' is only meaningful with "
+                                 "`average='samples'`, got `average='%s'`."
+                                 % average)
+            warn_message = ("'normalize' was deprecated in version 0.20 and "
+                            "will be removed in 0.22, instead use "
+                            "`average='%s'`." % average)
+            warnings.warn(warn_message, DeprecationWarning)
+
+        if not np.all(labels == present_labels):
+            if np.max(labels) > np.max(present_labels):
+                raise ValueError('All labels must be in [0, n, labels). '
+                                 'Got %d > %d' %
+                                 (np.max(labels), np.max(present_labels)))
+            if np.min(labels) < 0:
+                raise ValueError('All labels must be in [0, n, labels). '
+                                 'Got %d < 0' % np.min(labels))
+
+        if n_labels is not None:
+            y_true = y_true[:, labels[:n_labels]]
+            y_pred = y_pred[:, labels[:n_labels]]
+
         with np.errstate(divide='ignore', invalid='ignore'):
-            # oddly, we may get an "invalid" rather than a "divide" error here
-            pred_or_true = count_nonzero(y_true + y_pred, axis=1)
-            pred_and_true = count_nonzero(y_true.multiply(y_pred), axis=1)
+
+            if average == 'samples' or average == 'none-samples':
+                sum_axis = 1
+                class_weight = sample_weight
+                weights = None
+            elif average == 'micro':
+                sum_axis = 1
+                class_weight = None
+                weights = sample_weight
+            elif average == 'macro':
+                sum_axis = 0
+                class_weight = None
+                weights = sample_weight
+            elif average == 'weighted':
+                sum_axis = 0
+                class_weight = y_true.toarray().sum(axis=0)
+                weights = sample_weight
+            else:
+                sum_axis = 0
+                weights = sample_weight
+
+            pred_or_true = count_nonzero(y_true + y_pred, axis=sum_axis,
+                                         sample_weight=weights)
+            pred_and_true = count_nonzero(y_true.multiply(y_pred),
+                                          axis=sum_axis,
+                                          sample_weight=weights)
+            if average == 'micro':
+                pred_or_true = np.array([pred_or_true.sum()])
+                pred_and_true = np.array([pred_and_true.sum()])
+
             score = pred_and_true / pred_or_true
-            score[pred_or_true == 0.0] = 1.0
+            score[pred_or_true == 0.0] == 1.0
+
+            if average is not None:
+                if average == 'none-samples':
+                    score = np.sum(score)
```

Comment:
Oh I must have said something confusing. I don't think none-samples should give the sum. I don't think it is useful for anything to give the sum across samples, but I might be mistaken there. I intended that none-samples would return an array of jaccard scores for each sample, so the user can do whatever with them.

Reply:
If we go with returning an array, then we might need a deprecation, since the earlier behavior was to return this the sum of elements of array.

Reply:
We need to deprecate normalize in any case

Reply:
Done.